kad/kad.go

```diff
@@ -122,31 +100,21 @@ func Equal[K Key[K]](this, that NodeID[K]) bool {
 
 type Message interface{}
 
-type Request[K Key[K], A Address[A]] interface {
+type Request[K Key[K], N NodeID[K]] interface {
 	Message
 
 	// Target returns the target key and true, or false if no target key has been specfied.
 	Target() K
 
 	// EmptyResponse returns an empty response struct for this request message
 	// TODO: this is a weird patter, let's try to remove this.
-	EmptyResponse() Response[K, A]
+	EmptyResponse() Response[K, N]
 }
 
-type Response[K Key[K], A Address[A]] interface {
+type Response[K Key[K], N NodeID[K]] interface {
 	Message
 
-	CloserNodes() []NodeInfo[K, A]
-}
-
-type RoutingProtocol[K Key[K], N NodeID[K], A Address[A]] interface {
-	FindNode(ctx context.Context, to N, target K) (NodeInfo[K, A], []N, error)
-	Ping(ctx context.Context, to N) error
-}
-
-type RecordProtocol[K Key[K], N NodeID[K]] interface {
-	Get(ctx context.Context, to N, target K) ([]Record, []N, error)
-	Put(ctx context.Context, to N, record Record) error
+	CloserNodes() []N
```

Comment:
This is the part that seems like we would have to change it back one day. A useful Kademlia library needs to support getting addresses for nodes in the network.

Here's an alternative. 

Keep Request/Response, Address and NodeInfo.

Create a new type that just represents a set of nodes assumed to be close to a key:

```type KeyNodes[K Key[K], N NodeID[K]] struct {
    Key K
    Nodes []N
} 

Use this new type in all the routing and query state machines instead of `Request`. `Response`/`Request` can still be used in endpoint and simplequery.

Reply:
Another idea: let's rename `NodeID[K]` to `Node[K]` and let the implementation of `Node[K]` carry additional information. go-kademlia doesn't need to know anything about it.

Reply:
Suggestion from a private conversation.

Define 

```Go
type Node[K kad.Key, P any] interface {
  Key() K
  Preimage() P
}
```

So a `PeerNode` in libp2p is a `Node[key.Key256, peer.ID]`